consensus/istanbul/backend/announce.go

```diff
@@ -17,208 +17,386 @@
+	logger := sb.logger.New("func", "handleGetAnnouncesMsg")
+	var valAddresses []common.Address
 
-	// Send out an announce message when this thread starts
-	go sb.sendIstAnnounce()
+	if err := rlp.DecodeBytes(payload, &valAddresses); err != nil {
+		logger.Error("Error in decoding valAddresses", "err", err)
+		return err
+	}
 
-	// Set the initial states
-	announceThreadState := HighFreqBeforeFirstPeerState
-	currentTickerDuration := HighFreqTickerDuration
-	ticker := time.NewTicker(currentTickerDuration)
-	numSentMsgsInHighFreqAfterFirstPeerState := 0
+	sb.cachedAnnounceMsgsMu.RLock()
+	defer sb.cachedAnnounceMsgsMu.RUnlock()
+	// TODO:  Add support for the AnnounceMsg to contain multiple announce messages within it's payload
+	for _, valAddress := range valAddresses {
+		if cachedAnnounceMsgEntry, ok := sb.cachedAnnounceMsgs[valAddress]; ok {
+			logger.Trace("Sending announce msg", "peer", peer, "valAddress", valAddress)
+			go peer.Send(istanbulAnnounceMsg, cachedAnnounceMsgEntry.MsgPayload)
+		}
+	}
 
-	for {
-		select {
-		case <-sb.newEpochCh:
-			go sb.sendIstAnnounce()
+	return nil
+}
 
-		case <-ticker.C:
-			switch announceThreadState {
-			case HighFreqBeforeFirstPeerState:
-				{
-					if len(sb.broadcaster.FindPeers(nil, p2p.AnyPurpose)) > 0 {
-						announceThreadState = HighFreqAfterFirstPeerState
-					}
-				}
+// generateAndGossipAnnounce will generate the lastest announce msg from this node (as opposed to retrieving from the announceMsgCache) and then broadcast it to it's peers,
+// which should then gossip the announce msg message throughout the p2p network, since this announce msg's timestamp should be the latest among all of this
+// validator's previous announce msgs.
+func (sb *Backend) generateAndGossipAnnounce() error {
+	logger := sb.logger.New("func", "generateAndGossipAnnounce")
+	logger.Trace("generateAndGossipAnnounce called")
+	istMsg, announceVersion, announceAddress, err := sb.generateAnnounce()
+	if err != nil {
+		return err
+	}
 
-			case HighFreqAfterFirstPeerState:
-				{
-					if numSentMsgsInHighFreqAfterFirstPeerState >= 10 {
-						announceThreadState = LowFreqState
-					}
-					numSentMsgsInHighFreqAfterFirstPeerState += 1
-				}
+	if istMsg == nil {
+		return nil
+	}
 
-			case LowFreqState:
-				{
-					if currentTickerDuration != LowFreqTickerDuration {
-						// Reset the ticker
-						currentTickerDuration = LowFreqTickerDuration
-						ticker.Stop()
-						ticker = time.NewTicker(currentTickerDuration)
-					}
-				}
-			}
+	// Convert to payload
+	payload, err := istMsg.Payload()
+	if err != nil {
+		logger.Error("Error in converting Istanbul Announce Message to payload", "AnnounceMsg", istMsg.String(), "err", err)
+		return err
+	}
 
-			go sb.sendIstAnnounce()
+	// Add the generated announce message to this node's cache
+	sb.cachedAnnounceMsgs[announceAddress] = &announceMsgCachedEntry{MsgVersion: announceVersion, MsgPayload: payload}
```

Comment:
Because `cachedAnnounceMsgs` can be read/written outside of the main event loop we should be locking here

Reply:
Nice catch.